crates/holochain/src/core/workflow/call_zome_workflow.rs

```diff
@@ -111,44 +106,27 @@ async fn call_zome_workflow_inner<'env, Ribosome: RibosomeT>(
     let to_app_validate = {
         let workspace = workspace_lock.read().await;
         // Get the new head
-        let chain_head_end = workspace.source_chain.chain_head()?;
+        let chain_head_end_len = workspace.source_chain.len();
+        let new_elements_len = chain_head_end_len - chain_head_start_len;
 
         // collect all the elements we need to validate in wasm
-        let mut to_app_validate: Vec<Element> = vec![];
+        let mut to_app_validate: Vec<Element> = Vec::with_capacity(new_elements_len);
 
         // Has there been changes?
-        if chain_head_start != *chain_head_end {
-            // get the changes
-            let mut new_headers = workspace
-                .source_chain
-                .iter_back()
-                .scan(None, |current_header, element| {
-                    let my_header = current_header.clone();
-                    *current_header = element.header().prev_header().cloned();
-                    let r = match my_header {
-                        Some(current_header) if current_header == chain_head_start => None,
-                        _ => Some(element),
-                    };
-                    Ok(r)
-                })
-                .map_err(WorkflowError::from);
-
-            while let Some(header) = new_headers.next()? {
-                let chain_element = workspace
-                    .source_chain
-                    .get_element(header.header_address())?;
-                let prev_chain_element = match chain_element {
-                    Some(ref c) => match c.header().prev_header() {
-                        Some(h) => workspace.source_chain.get_element(&h)?,
-                        None => None,
-                    },
-                    None => None,
-                };
-                if let Some(ref chain_element) = chain_element {
-                    sys_validate_element(&agent_key, chain_element, prev_chain_element.as_ref())
-                        .await?;
-                    to_app_validate.push(chain_element.to_owned());
-                }
+        if new_elements_len > 0 {
+            // Loop forwards through all the new elements
+            let mut i = chain_head_start_len;
+            while let Some(element) = workspace.source_chain.get_at_index(i as u32)? {
+                // TODO: Figure out how to write the cache from
+                // this validation call as there's likely some gets
```

Comment:
Not sure how to do this without also writing out the call zomes data prematurely. This also wouldn't be an issue if we had an authored db